__Задача 1__

Дана матрица объект-признак
```
X = [[ 1.22401313,  2.30868478,  3.03636353,  2.69287214],
     [-0.18757272,  1.30337355,  5.12093014,  3.46363202],
     [-0.81094525,  1.82463398,  5.79686488,  1.86159445],
     [ 0.75129018,  2.67392052,  3.65529809,  1.66746094],
     [ 0.00972362,  1.97367255,  2.50594319,  1.69755173],
     [-0.62972637,  0.77750764,  2.84124027,  4.54410559],
     [ 2.29536229,  1.81206697,  1.95026215,  1.51874636],
     [ 0.0920418 ,  2.26971361,  7.47708735,  2.61081203],
     [ 2.39252799,  3.17563985,  3.61420599,  5.10773362],
     [ 0.54983815,  2.87988651,  1.65752765,  1.59635987]]
```
и значения целевой переменной
```
y = [ 9.26193358,  9.700363  ,  8.67214805,  8.74796974,  6.18689108,
      7.53312713,  7.57643777, 12.44965478, 14.29010746,  6.68361218]
```

Подберите два признака (из четырёх) так, чтобы уровень линейной зависимости целевой переменной от значений этих признаков был максимальным. Другими словами, модель линейной регрессии на этих признаках должна давать наилучший результат. В качестве ответа запишите значение коэффициента детерминации для модели на этих признаках.

In [199]:
import numpy as np
from sklearn.linear_model import LinearRegression
import pandas as pd
from operator import itemgetter
from scipy import stats

In [200]:
x = np.array([[ 1.22401313,  2.30868478,  3.03636353,  2.69287214],
     [-0.18757272,  1.30337355,  5.12093014,  3.46363202],
     [-0.81094525,  1.82463398,  5.79686488,  1.86159445],
     [ 0.75129018,  2.67392052,  3.65529809,  1.66746094],
     [ 0.00972362,  1.97367255,  2.50594319,  1.69755173],
     [-0.62972637,  0.77750764,  2.84124027,  4.54410559],
     [ 2.29536229,  1.81206697,  1.95026215,  1.51874636],
     [ 0.0920418 ,  2.26971361,  7.47708735,  2.61081203],
     [ 2.39252799,  3.17563985,  3.61420599,  5.10773362],
     [ 0.54983815,  2.87988651,  1.65752765,  1.59635987]])

In [201]:
y = np.array([ 9.26193358,  9.700363  ,  8.67214805,  8.74796974,  6.18689108,
      7.53312713,  7.57643777, 12.44965478, 14.29010746,  6.68361218])

In [202]:
df = pd.DataFrame()
df[0] = np.ones(x.shape[0])
for i in range(1, 5):
  df[i] = np.array(x.transpose()[i-1])
df[6] = y
df

0         1         2         3         4          6
0  1.0  1.224013  2.308685  3.036364  2.692872   9.261934
1  1.0 -0.187573  1.303374  5.120930  3.463632   9.700363
2  1.0 -0.810945  1.824634  5.796865  1.861594   8.672148
3  1.0  0.751290  2.673921  3.655298  1.667461   8.747970
4  1.0  0.009724  1.973673  2.505943  1.697552   6.186891
5  1.0 -0.629726  0.777508  2.841240  4.544106   7.533127
6  1.0  2.295362  1.812067  1.950262  1.518746   7.576438
7  1.0  0.092042  2.269714  7.477087  2.610812  12.449655
8  1.0  2.392528  3.175640  3.614206  5.107734  14.290107
9  1.0  0.549838  2.879887  1.657528  1.596360   6.683612

In [203]:
def sum_of_squares(samples: np.ndarray) -> float:
    """Сумма квадратов отклонений.
    """
    return ((samples - samples.mean()) ** 2).sum()

In [204]:
feach_numb = [1, 2, 3, 4] 
feach_r2_list = []
y = df[6].values
for i in feach_numb:
  feach_numb = [1, 2, 3, 4] 
  feach_numb.remove(i)
  for j in feach_numb:
    x = df[[0, i, j]].values
    lr = LinearRegression(fit_intercept=False).fit(x, y)
    z = x.dot(lr.coef_)
    e = y - z
    R = 1 - sum_of_squares(e) / sum_of_squares(y)
    feach_r2_list.append((i, j, R))

feach_r2_list

[(1, 2, 0.18113594742585226),
 (1, 3, 0.7634246238793152),
 (1, 4, 0.4532966783144077),
 (2, 1, 0.18113594742585204),
 (2, 3, 0.5479482734039008),
 (2, 4, 0.6062055761129932),
 (3, 1, 0.7634246238793151),
 (3, 2, 0.5479482734039011),
 (3, 4, 0.622441987650532),
 (4, 1, 0.4532966783144078),
 (4, 2, 0.6062055761129932),
 (4, 3, 0.622441987650532)]

In [205]:
max(feach_r2_list,key=itemgetter(2))

(1, 3, 0.7634246238793152)

Таким образом, модель линейной регрессии на первом и третьем признаках дает наилучший результат.

__Задача 2__

Является ли значимым уравнение регрессии, полученное в предыдущей задаче? В качестве ответа запишите 1, если является, и 0 иначе.

In [206]:
alpha = 0.05
n = x.shape[0]
m = x.shape[1] - 1

k1 = m
k2 = n - m - 1

k1, k2

(2, 7)

In [207]:
t = stats.f.ppf(1 - alpha, k1, k2)
t

4.73741412777588

In [208]:
R = max(feach_r2_list,key=itemgetter(2))[2]
F = (R / k1) / ((1 - R) / k2)
F

11.29443912292265

11.29443912292265 > 4.73741412777588. 

Статистика попала в критическую область.

Следовательно, уравнение регрессии также признаётся значимым.

__Задача 3__

Для проведения A/B-тестирования сайта интернет-магазина были получены следующие данные: страница `A` была посещена 2509 раз, из них 77 закончились совершением покупки, страница `B` была посещена 1465 раз, 60 из них закончились совершением покупки. Является ли значимым отличие конверсии на страницах `A` и `B`? В качестве ответа запишите 1, если является, и 0 иначе.

_Подсказка_. Реализуйте двухвыборочный t-тест. В качестве выборок здесь можно взять наборы меток совершения покупки (0 или 1) каждым посетителем.

In [209]:
n1 = 2509
n2 = 1465

In [210]:
X_A = np.array([1 for _ in range(77)] + [0 for _ in range(n1 - 77)])
X_B = np.array([1 for _ in range(60)] + [0 for _ in range(n2 - 77)])

In [217]:
s1 = X_A.std(ddof=1)
s2 = X_B.std(ddof=1)
df = (s1 ** 2 / n1 + s2 ** 2 / n2) ** 2 / \
    ((s1 ** 2 / n1) ** 2 / (n1 - 1) + (s2 ** 2 / n2) ** 2 / (n2 - 1))
alpha = 0.05

t1 = stats.t.ppf(alpha / 2, df=df)
t2 = stats.t.ppf(1 - alpha / 2, df=df)

t1, t2

(-1.9608364067753534, 1.960836406775353)

In [212]:
stats.ttest_ind(X_A, X_B, equal_var=False)

Ttest_indResult(statistic=-1.7140755319265335, pvalue=0.08663052234426083)

За нулевую гипотезу берётся предположение, что конверсия на страницах A и B не отличается.

р-значение выше уровня значимости по-умолчанию = 0.05, значение статистики не попало в критическую область.

Следовательно, отличие конверсии на страницах `A` и `B` является незначимым.